casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala

```diff
@@ -239,73 +239,80 @@ object BlockAPI {
     }
   }
 
-  def visualizeDag[
-      F[_]: Monad: Sync: MultiParentCasperRef: Log: SafetyOracle: BlockStore,
-      G[_]: Monad: GraphSerializer
-  ](
-      d: Option[Int] = None,
-      visualizer: (Vector[Vector[BlockHash]], String) => F[G[Graphz[G]]],
-      stringify: G[Graphz[G]] => String
-  ): Effect[F, VisualizeBlocksResponse] = {
+  private def toposortDag[
+      F[_]: Monad: MultiParentCasperRef: Log: SafetyOracle: BlockStore,
+      A
+  ](maybeDepth: Option[Int])(
+      doIt: (MultiParentCasper[F], Vector[Vector[BlockHash]]) => F[ApiErr[A]]
+  ): Effect[F, A] = {
 
     val errorMessage =
       "Could not visualize graph, casper instance was not available yet."
 
-    def casperResponse(implicit casper: MultiParentCasper[F]): Effect[F, VisualizeBlocksResponse] =
+    def casperResponse(implicit casper: MultiParentCasper[F]): Effect[F, A] =
       for {
-        dag                <- MultiParentCasper[F].blockDag
-        maxHeight          <- dag.topoSort(0L).map(_.length - 1)
-        depth              = d.getOrElse(maxHeight)
-        startHeight        = math.max(0, maxHeight - depth)
-        topoSort           <- dag.topoSortTail(depth)
-        lastFinalizedBlock <- MultiParentCasper[F].lastFinalizedBlock
-        graph              <- visualizer(topoSort, PrettyPrinter.buildString(lastFinalizedBlock.blockHash))
-      } yield VisualizeBlocksResponse(stringify(graph)).asRight
-
-    MultiParentCasperRef.withCasper[F, ApiErr[VisualizeBlocksResponse]](
+        dag       <- MultiParentCasper[F].blockDag
+        maxHeight <- dag.topoSort(0L).map(_.length - 1)
+        depth     = maybeDepth.getOrElse(maxHeight)
```

Comment:
`dag.topoSort` sounds expensive and you shouldn't be doing it unless maybeDepth is None.

Reply:
👍

Reply:
We should not be calling topoSort for finding out which block is the highest anyway (this is a ridicilous overkill). I have ticketed this https://rchain.atlassian.net/browse/RCHAIN-3253

Reply:
Yes, I've even added TODO for that. Good there is a ticket. But for now we need to know the depth. This part of code is just extracting what was common in two methods already (`showBlocks` and `vdag`) and was about to hit the third method (`machineVerifiableDag`)